File: sdk/python/base_connector.py
"""
Base Connector Class for Connector Platform SDK
"""
from typing import Dict, Optional, Any
import requests


class BaseConnector:
    """
    Base class for all connectors in the platform.
    Provides common functionality for API calls and authentication.
    """
    
    def __init__(self, connection_id: str, config: Dict):
        self.connection_id = connection_id
        self.config = config
        self.connector_name = None
        self.base_url = None
        self.platform_url = config.get("platform_url", "http://localhost:5000")
    
    def execute_request(
        self,
        endpoint_config: Dict,
        params: Optional[Dict] = None,
        body: Optional[Dict] = None,
        path_params: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Execute an API request through the platform proxy.
        
        Args:
            endpoint_config: Configuration for the endpoint (method, path, headers)
            params: Query parameters
            body: Request body for POST/PUT/PATCH requests
            path_params: Parameters to substitute in the path
        
        Returns:
            Dict containing the API response
        """
        proxy_url = f"{self.platform_url}/api/v1/proxy/execute"
        
        payload = {
            "connection_id": self.connection_id,
            "endpoint_config": endpoint_config,
            "params": params,
            "body": body,
            "path_params": path_params
        }
        
        try:
            response = requests.post(proxy_url, json=payload, timeout=30)
            return response.json()
        except requests.exceptions.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    def get_connection_info(self) -> Dict[str, Any]:
        """Get information about the current connection."""
        url = f"{self.platform_url}/api/v1/connections/{self.connection_id}"
        
        try:
            response = requests.get(url, timeout=10)
            return response.json()
        except requests.exceptions.RequestException as e:
            return {
                "success": False,
                "error": str(e)
            }

File: sdk/go/connector/base_connector.go
package connector

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type BaseConnector struct {
	ConnectionID  string
	Config        map[string]interface{}
	ConnectorName string
	BaseURL       string
	PlatformURL   string
}

type ProxyRequest struct {
	ConnectionID   string                 `json:"connection_id"`
	EndpointConfig map[string]interface{} `json:"endpoint_config"`
	Params         map[string]interface{} `json:"params,omitempty"`
	Body           map[string]interface{} `json:"body,omitempty"`
	PathParams     map[string]interface{} `json:"path_params,omitempty"`
}

type APIResponse struct {
	Success    bool                   `json:"success"`
	StatusCode int                    `json:"status_code,omitempty"`
	Data       map[string]interface{} `json:"data,omitempty"`
	Error      string                 `json:"error,omitempty"`
}

func NewBaseConnector(connectionID string, config map[string]interface{}) *BaseConnector {
	platformURL, ok := config["platform_url"].(string)
	if !ok {
		platformURL = "http://localhost:5000"
	}

	return &BaseConnector{
		ConnectionID: connectionID,
		Config:       config,
		PlatformURL:  platformURL,
	}
}

func (bc *BaseConnector) ExecuteRequest(
	endpointConfig map[string]interface{},
	params map[string]interface{},
	body map[string]interface{},
	pathParams map[string]interface{},
) (map[string]interface{}, error) {
	proxyURL := fmt.Sprintf("%s/api/v1/proxy/execute", bc.PlatformURL)

	proxyReq := ProxyRequest{
		ConnectionID:   bc.ConnectionID,
		EndpointConfig: endpointConfig,
		Params:         params,
		Body:           body,
		PathParams:     pathParams,
	}

	jsonData, err := json.Marshal(proxyReq)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Post(proxyURL, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var apiResp APIResponse
	if err := json.Unmarshal(bodyBytes, &apiResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	if !apiResp.Success {
		return nil, fmt.Errorf("API error: %s", apiResp.Error)
	}

	return apiResp.Data, nil
}

func (bc *BaseConnector) GetConnectionInfo() (map[string]interface{}, error) {
	url := fmt.Sprintf("%s/api/v1/connections/%s", bc.PlatformURL, bc.ConnectionID)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return result, nil
}
